File: Insertion_Sort/Objective-C/InsertionSort.m
//
//  InsertionSort.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "InsertionSort.h"

@implementation InsertionSort

+ (void)sort:(NSMutableArray<NSNumber *> *)array order:(InsertionSortOrder)order {
    NSUInteger arrayCount = array.count;

    if (arrayCount < 2) {
        return;
    }

    for (NSUInteger i = 1; i < arrayCount; i++) {
        NSNumber *currentNumber = array[i];

        for (NSUInteger j = i; j > 0; j--) {
            NSUInteger prevIndex = j - 1;
            NSNumber *prevNumber = array[prevIndex];
            NSComparisonResult comparisonResult = [prevNumber compare:currentNumber];

            if ((order == InsertionSortOrderAsc && comparisonResult == NSOrderedDescending) ||
                (order == InsertionSortOrderDesc && comparisonResult == NSOrderedAscending)) {
                [array exchangeObjectAtIndex:prevIndex withObjectAtIndex:j];
            }
            else {
                break;
            }
        }
    }
}

@end

File: Insertion_Sort/Swift/InsertionSort.swift
//
//  InsertionSort.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct InsertionSort {

    enum SortOrder {
        case ASC
        case DESC
    }

    /** Complexity is O(n^2) */
    static public func sort<T: Comparable>(_ array: [T?]?, order: SortOrder) -> [T?]? {
        let arrayCount = array?.count ?? 0

        if arrayCount < 2 {
            return array
        }

        var resultArray = array!

        for i in 1..<arrayCount {
            let currentValue = resultArray[i]

            for j in stride(from: i, to: 0, by: -1) {
                let prevIndex = j - 1
                let prevValue = resultArray[prevIndex]

                if (order == .ASC && ((prevValue != nil && currentValue == nil) || (prevValue != nil && currentValue != nil && prevValue! > currentValue!))) ||
                    (order == .DESC && ((prevValue == nil && currentValue != nil) || (prevValue != nil && currentValue != nil && prevValue! < currentValue!))) {
                    swap(&resultArray[prevIndex], &resultArray[j])
                }
            }
        }

        return resultArray;
    }

}
